src/commands/up.rs

```diff
@@ -646,6 +667,87 @@ fn trigger_command_for_resolved_app_source(
+    let tmp_app = spin_app::App::new("tmp", locked_app.clone());
+    let component_trigger_id_map = component_trigger_map(&tmp_app, components)?;
+    let components = component_trigger_id_map.keys().collect::<HashSet<_>>();
+    let triggers = component_trigger_id_map
+        .values()
+        .flatten()
+        .collect::<HashSet<_>>();
+    locked_app.components.retain(|c| components.contains(&c.id));
+    locked_app.triggers.retain(|t| triggers.contains(&t.id));
+    Ok(())
+}
+
+// Given a list of components and App, returns a map of component ID to trigger ID.
+// Also validates that the components exist in the app and have valid service chaining.
+fn component_trigger_map(
+    app: &spin_app::App,
+    components: &[String],
+) -> Result<ComponentTriggerMap> {
+    let component_triggers: Vec<(AppComponent<'_>, String)> = app
+        .triggers()
+        .filter_map(|t| match t.component() {
+            Ok(comp) => {
+                if components.contains(&comp.id().to_string()) {
+                    Some((comp, t.id().to_string()))
+                } else {
+                    None
+                }
+            }
+            Err(_) => None,
+        })
+        .collect();
+
+    // Allow only service chaining among retained components
+    // All wildcard service chaining is disallowed
+    for (c, _) in &component_triggers {
+        let allowed_hosts = allowed_hosts(c)?;
+        allowed_hosts.iter().try_for_each(|host| {
+            let uri = host.parse::<http::Uri>().unwrap();
+            if let Some(local_component) = parse_service_chaining_target(&uri) {
+                if !components.contains(&local_component) {
+                    if local_component == "*" {
+                        bail!("Component selected with '--component {}' cannot use wildcard service chaining:
+  allowed_outbound_hosts = [\"http://*.spin.internal\"]", c.id());
+                    }
+                    bail!(
+                        "Component selected with '--component {}' cannot use service chaining to unselected component:
+  allowed_outbound_hosts = [\"http://{}.spin.internal\"]",
+                        c.id(), local_component
+                    );
+                }
+            }
+            Ok(())
+        })?;
+    }
+    let mut component_trigger_map: ComponentTriggerMap = HashMap::new();
+    component_triggers.into_iter().for_each(|(c, t)| {
+        component_trigger_map
+            .entry(c.id().to_string())
+            .or_insert_with(|| HashSet::from([t]));
+    });
+
+    // Ensure that the requested components exist in the app
+    let mut not_found = Vec::new();
+    components.iter().for_each(|c| {
+        if !component_trigger_map.contains_key(c) {
+            not_found.push(c);
+        }
+    });
+    ensure!(
+        not_found.is_empty(),
+        "Specified components not found in application: {:?}",
+        not_found
+    );
+
+    Ok(component_trigger_map)
```

Comment:
I got quite lost in this function I'm afraid.  Conceptually the idea seems to be "remove all the components that aren't in the list, then remove all the triggers whose components are no longer present."  (Or, equivalently, "create two remove lists, one of all the components that aren't in the list, and one of all the triggers whose components aren't in the list.")  I didn't quite get where the mapping came in, or why the service chaining validation and UI sanity checking were intermingled.  If this is the best way to do it then fair enough, it just seemed like it could be simpler but I could well be missing a lot of nuance or edge cases!

Reply:
I agree that it takes on two purposes: 
1. Get a mapping of component ID to trigger ID
2. Ensure that component retention request is valid

I clumped it together in order to reduce iterations but i think if i make the component trigger map hold the `AppComponent` then i could make step 2 a subsequent step

Reply:
Reorganized the code to make this cleaner and clearer. Let me know what you think